localstack/services/s3/notifications.py

```diff
@@ -0,0 +1,513 @@
+
+
+class Notification(TypedDict):
+    Records: List[EventRecord]
+
+
+class S3EventNotificationContext:
+    def __init__(self, context: RequestContext):
+        self.event_type = EVENT_OPERATION_MAP.get(context.operation.wire_name, "")
+        self.region = context.region
+        self.bucket_name = context.service_request["Bucket"]
+        self.key_name = context.service_request["Key"]
+        self.xray = context.request.headers.get(HEADER_AMZN_XRAY)
+
+        moto_backend = get_moto_s3_backend(context)
+        bucket: FakeBucket = get_bucket_from_moto(moto_backend, bucket=self.bucket_name)
+        key: FakeKey = get_key_from_moto_bucket(moto_bucket=bucket, key=self.key_name)
+        self.bucket_location = bucket.location
+        self.key_size = key.contentsize
+        self.key_etag = key.etag.strip('"')
+        self.key_name = quote(key.name)
+        self.key_version_id = key.version_id if bucket.is_versioned else None  # todo: check this?
+
+
+def _matching_event(events: EventList, event_name: str) -> bool:
+    """
+    See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html
+    Checks if the event is part of the NotificationConfiguration, and returns if the event should be notified for
+    this configuration
+    :param events: the list of events of the NotificationConfiguration
+    :param event_name: the event type, like s3:ObjectCreated:* or s3:ObjectRemoved:Delete
+    :return: boolean indicating if the event should be sent to the notifiers
+    """
+    if event_name in events:
+        return True
+    wildcard_pattern = f"{event_name[0:event_name.rindex(':')]}:*"
+    return wildcard_pattern in events
+
+
+def _matching_filter(
+    notification_filter: Optional[NotificationConfigurationFilter], key_name: str
+) -> bool:
+    """
+    See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-filtering.html
+    S3 allows filtering notification events with rules about the key name.
+    If the key does not have a filter rule, or if it matches the rule, then returns that the event should be sent
+    :param notification_filter: the Filter structure from NotificationConfiguration
+    :param key_name: the key name of the key concerned by the event
+    :return: boolean indicating if the key name matches the rules and the event should be sent
+    """
+    # TODO: implement wildcard filtering
+    if not notification_filter or not notification_filter.get("Key", {}).get("FilterRules"):
+        return True
+    filter_rules = notification_filter.get("Key").get("FilterRules")
+    for rule in filter_rules:
+        name = rule.get("Name", "").lower()
+        value = rule.get("Value", "")
+        if name == "prefix" and not key_name.startswith(value):
+            return False
+        elif name == "suffix" and not key_name.endswith(value):
+            return False
+
+    return True
+
+
+class BaseNotifier:
+    service_name = None
```

Comment:
```suggestion
    service_name: str
```
i think it's better to declare it but not set it, that way you'll get IDE warnings if you don't set the service name in the implementing service